add desk publisher switcher UIP

tinnus-napbus opened the pull request:
This UIP is for the ability for app publishers to tell subscribers to switch their source of updates to a new ship/desk. This is useful for app developers who need to migrate where they're distributing their app from.

belisarius222 commented:
@joemfb Security: it would be safer to prompt the user to allow the source to switch.  Maybe I trust you but I don't trust your friends.  The UI should let me decide whether I should accept it or not.

@philipcmonk So many people would drop off if we prompt that it doesn't seem worth it.

@ashelkovnykov It seems like this policy should not be the responsibility of the kernel.  Different userspace app distribution systems might want to have different policies.

@frodwith Instead of performing the source switch itself, Kiln could emit a subscription notification that the developer wants the user to switch their distribution source.  Each app distribution platform (e.g. Tlon's Garden, Holium's Realm) could make its own decision about how to handle this notification.  One might automatically perform the switch; another might perform the switch automatically but also notify the user that the switch had happened; another could prompt the user about whether to perform the switch.

This last proposal had wide buy-in among the core devs.  @tinnus-napbus When you get a chance, I recommend revising this UIP accordingly.

tinnus-napbus commented:
I have changed it so requests are stored in kiln's state and users have to approve them. I have added %watch and scry paths with json conversion methods so front-ends can prompt the user to approve. See the updated UIP for more details & the updated branch linked therein for the implementation.
